token-metadata/program/src/processor/collection/unverify_collection.rs

```diff
@@ -1,71 +1,122 @@
+    let collection_metadata_info = next_account_info(account_info_iter)?;
     let edition_account_info = next_account_info(account_info_iter)?;
+
     let using_delegated_collection_authority = accounts.len() == 6;
 
-    assert_signer(collection_authority_info)?;
+    // Account validation.
     assert_owned_by(metadata_info, program_id)?;
-    assert_owned_by(collection_info, program_id)?;
-    assert_owned_by(collection_mint, &spl_token::id())?;
-    assert_owned_by(edition_account_info, program_id)?;
+    assert_signer(collection_authority_info)?;
+    assert_owned_by(collection_mint_info, &spl_token::id())?;
 
+    // Deserialize the collection item metadata.
     let mut metadata = Metadata::from_account_info(metadata_info)?;
-    let collection_data = Metadata::from_account_info(collection_info)?;
-
-    assert_collection_verify_is_valid(
-        &metadata.collection,
-        &collection_data,
-        collection_mint,
-        edition_account_info,
-    )?;
+
+    // First, if there's no collection set, we can just short-circuit
+    // since there's nothing to unverify.
+    if metadata.collection.is_none() {
+        return Ok(());
+    }
+
+    // Unwrap ok because of check above.
+    let collection = metadata.collection.clone().unwrap();
+
+    // Short-circuit if it's already unverified.
+    if !collection.verified {
+        return Ok(());
+    }
+
+    // The collection parent must be the actual parent of the
+    // collection item.
+    if collection.key != *collection_mint_info.key {
+        return Err(MetadataError::NotAMemberOfCollection.into());
+    }
+
+    // We need to ensure the metadata is derived from the mint so
+    // someone cannot pass in a burned metadata account associated with
+    // a different mint.
+    assert_metadata_derivation(program_id, collection_metadata_info, collection_mint_info)?;
+
+    // Check if the collection metadata account is burned. If it is,
+    // there's no sized data to update and the user can simply unverify
+    // the NFT.
+    //
+    // This check needs to happen before the program owned check.
+    let parent_burned = collection_metadata_info.data_is_empty()
+        && collection_metadata_info.owner == &system_program::ID;
+
+    if parent_burned {
+        // If the parent is burned, we need to check that the authority
+        // is the update authority on the item metadata and then we can
+        // just unverify the NFT and return.
+        //
+        // Collection Delegates for burned collection parents should not be
+        // respected as there's currently no way to revoke them.
+
+        if metadata.update_authority != *collection_authority_info.key {
+            return Err(MetadataError::UpdateAuthorityIncorrect.into());
+        }
+
+        metadata.collection.as_mut().unwrap().verified = false;
+        clean_write_metadata(&mut metadata, metadata_info)?;
+        return Ok(());
+    }
```

Comment:
Style nit:  Instead of setting `collection.verified = false` and `clean write metadata` in this block and then returning early, I would actually put the rest of this code in an else statement, except have the `collection.verified = false` and `clean write metadata` at the end, outside of the if-else block.  To me that is slightly clearer control flow.  But this is a style nit based on my opinion so feel free to leave it as-is.